src/CustomElementInternals.js

```diff
@@ -0,0 +1,294 @@
+import CustomElementDefinition from './CustomElementDefinition';
+import * as Utilities from './Utilities';
+import * as CESymbols from './CustomElementInternalSymbols';
+import CEState from './CustomElementState';
+
+export default class CustomElementInternals {
+  constructor() {
+    /** @type {!Map<string, !CustomElementDefinition>} */
+    this._localNameToDefinition = new Map();
+
+    /** @type {!Map<!Function, !CustomElementDefinition>} */
+    this._constructorToDefinition = new Map();
+
+    /** @type {!Array<!function(!Node)>} */
+    this._patches = [];
+  }
+
+  /**
+   * @param {string} localName
+   * @param {!CustomElementDefinition} definition
+   */
+  setDefinition(localName, definition) {
+    this._localNameToDefinition.set(localName, definition);
+    this._constructorToDefinition.set(definition.constructor, definition);
+  }
+
+  /**
+   * @param {string} localName
+   * @return {!CustomElementDefinition|undefined}
+   */
+  localNameToDefinition(localName) {
+    return this._localNameToDefinition.get(localName);
+  }
+
+  /**
+   * @param {!Function} constructor
+   * @return {!CustomElementDefinition|undefined}
+   */
+  constructorToDefinition(constructor) {
+    return this._constructorToDefinition.get(constructor);
+  }
+
+  /**
+   * @param {!function(!Node)} listener
+   */
+  addPatch(listener) {
+    this._patches.push(listener);
+  }
+
+  /**
+   * @param {!Node} node
+   */
+  patchTree(node) {
+    Utilities.walkDeepDescendants(node, node => this.patch(node));
+  }
+
+  /**
+   * @param {!Node} node
+   */
+  patch(node) {
+    if (node[CESymbols.patched]) return;
```

Comment:
I've added a flag that gets flipped on the first call to `addPatch`, so both `patchTree` and `patch` should bail if there's nothing to do. I'd prefer not to have to put all the patches in one place; ideally they would be grouped with all the other (non-instance) patches that are being applied to a particular prototype. (`innerHTML` defined in the same place as other patches for Element and `textContent` with Node, if it ends up being necessary.)